drivers: add cc110x driver

kaspar030 opened the pull request:
This PR adds a cc110x driver that is based on netdev2 + gnrc.

Tested on msba2. Please note that the wireless parameters are different than with our legacy code, so communication with devices using the old drivers will probably not work.

The gnrc glue code fakes a iee802154 device so sixlowpan can be used in cc110x networks.

Waiting for #3683.

kaspar030 commented:
- rebased

kaspar030 commented:
- rebased

OlegHahm commented on boards/msba2/include/cc110x_params.h:
```diff
@@ -0,0 +1,43 @@
+ */
+const cc110x_params_t cc110x_params[] = {
+    { .spi=0,
+        .CS=53,
```
Sorry for the nitpick, bu tthe indentation seems off.

OlegHahm commented on drivers/cc110x/cc110x-defaultsettings.c:
```diff
@@ -0,0 +1,98 @@
+              0x8E,         ///<   0 dBm
+              0x85,         ///< + 5 dBm
+              0xCC,         ///< + 7 dBm
+              0xC3          ///< +10 dBm
```
While you're at it you could fix the comment style.

OlegHahm commented on drivers/cc110x/cc110x-defaultsettings.c:
```diff
@@ -0,0 +1,98 @@
+const char cc110x_default_conf[] = {
+    0x06, /* IOCFG2 */
+    0x2E, /* IOCFG1 */
+#ifdef CC110X_IOCONF0_VAL
```
What's that?

kaspar030 replied:
One board I had to do a port for uses the cc110x as clock source, so SPI&cc1101 get initialized in the board config. IOCONF0 is set to output a clock. If later at driver initialization the cc110x gets reset or ioconf0 overwritten, the board hangs.

I'll add a comment.

OlegHahm commented on drivers/cc110x/cc110x-netdev2.c:
```diff
@@ -0,0 +1,270 @@
+
+#else
+    for (int i = 0; i < 8; i++) {
+        eui64[i] = i;
```
Wouldn't some pseudo random number make more sense?

kaspar030 replied:
probably, will change.

OlegHahm commented:
Which packet size does this driver support?

OlegHahm commented on drivers/cc110x/cc110x-netdev2.c:
```diff
@@ -0,0 +1,270 @@
+
+static int _set(netdev2_t *dev, netopt_t opt, void *value, size_t value_len)
+{
+    DEBUG("%s:%u\n", __func__, __LINE__);
```
Why do we have this here but not at similar places?

kaspar030 replied:
leftover, will remove.

OlegHahm commented on drivers/cc110x/cc110x-netdev2.c:
```diff
@@ -0,0 +1,270 @@
+    return 0;
+}
+
+#if 0
```
Hm?

kaspar030 replied:
ups

OlegHahm commented on drivers/cc110x/cc110x-rxtx.c:
```diff
@@ -0,0 +1,294 @@
+
+static void _rx_start(cc110x_t *dev)
+{
+//    DEBUG("%s:%s:%u\n", RIOT_FILE_RELATIVE, __func__, __LINE__);
```
remove

OlegHahm commented on drivers/cc110x/cc110x-rxtx.c:
```diff
@@ -0,0 +1,294 @@
+        /* set GDO2 to 0x2 -> will deassert at TX FIFO below threshold */
+        gpio_irq_enable(dev->params.GDO2);
+        cc110x_write_reg(dev, CC110X_IOCFG2, 0x02);
+    } else {
```
coding conventions

OlegHahm commented on drivers/cc110x/cc110x-rxtx.c:
```diff
@@ -0,0 +1,294 @@
+        case RADIO_TX_BUSY:
+            if (!gpio_read(dev->params.GDO2)) {
+                _tx_continue(dev);
+            } else {
```
.

OlegHahm commented on drivers/cc110x/cc110x-spi.c:
```diff
@@ -0,0 +1,175 @@
+void cc110x_cs(cc110x_t *dev)
+{
+    volatile int retry_count = 0;
+//    /* Switch MISO/GDO1 to GPIO input mode */
```
double comment

OlegHahm commented on drivers/cc110x/cc110x-spi.c:
```diff
@@ -0,0 +1,175 @@
+        }
+    }
+    /* Switch MISO/GDO1 to spi mode */
+    //spi_conf_pins(dev->params.spi);
```
remove

kaspar030 replied:
oh, actually this shouldn't work with setting to gpio above, not setting to spi here. wierd, this code does run on msba2. Have to check why. ;)

OlegHahm commented:
Sorry for the superficial review so far - will test asap. Well documented though.

OlegHahm commented:
Could you add the avsextrem (and maybe pttu) as well?

OlegHahm commented on drivers/include/cc110x/cc110x-internal.h:
```diff
@@ -0,0 +1,202 @@
+ */
+typedef struct __attribute__((packed))
+{
+    uint8_t length;                         ///< Length of the packet (without length byte)
```
C style comments here, too.

OlegHahm commented on drivers/include/cc110x/cc110x-internal.h:
```diff
@@ -0,0 +1,202 @@
+| Length | Address | PhySrc  | Flags |    Data    |
+|        |         |         |       |            |
+---------------------------------------------------
+  1 byte   1 byte    1 byte   1 byte   <= 58 bytes
```
I guess the payload size isn't true any more!?

kaspar030 replied:
true

OlegHahm commented on drivers/include/cc110x/cc110x-interface.h:
```diff
@@ -0,0 +1,62 @@
+extern "C" {
+#endif
+
+char *cc110x_get_marc_state(cc110x_t *dev);
```
doxygen will complain about this file...

OlegHahm commented on drivers/include/cc110x/cc110x-internal.h:
```diff
@@ -0,0 +1,202 @@
+#define CC110X_GDO1_LOW_RETRY       (100)   /**< Max. retries for SO to go low
+                                                 after CS */
+#define CC110X_DEFAULT_CHANNEL      (0)     /**< The default channel number */
+#define CC110X_MIN_CHANNR           (0)     /**< lowest possible channel number */
```
0 is a valid channel?

kaspar030 replied:
yep

OlegHahm commented on drivers/include/cc110x/cc110x-defines.h:
```diff
@@ -0,0 +1,207 @@
+ * If variable packet length is configured in PKTCTRL0 the
+ * first byte after the synch word determines the packet length.
+ */
+#define VARIABLE_PKTLEN     (0x01)
```
Where is this used?

kaspar030 replied:
nowhere, I copied the defines from the old code. should I remove all unused? maybe we use them later...

OlegHahm replied:
doesn't hurt to let them in for now

OlegHahm commented on drivers/include/cc110x/cc110x-defines.h:
```diff
@@ -0,0 +1,207 @@
+#define I_RSSI              (0x00)      ///< Index 0 contains RSSI information (from optionally appended packet status bytes).
+#define I_LQI               (0x01)      ///< Index 1 contains LQI & CRC_OK information (from optionally appended packet status bytes).
+#define MARC_STATE          (0x1F)      ///< Bitmask (=00011111) for reading MARC_STATE in MARCSTATE status register.
+#define PKTSTATUS_CS                  (0x40)      ///< Bitmask (=01000000) for reading CS (Carrier Sense) in PKTSTATUS status register.
```
indentation is off

haukepetersen replied:
and no c-style comments

OlegHahm commented on drivers/cc110x/cc110x.c:
```diff
@@ -0,0 +1,393 @@
+    dev->radio_state = RADIO_IDLE;
+}
+
+char *cc110x_get_marc_state(cc110x_t *dev)
```
Can't we make this and the following `to_str()` functions `DEVELHELP` only?

haukepetersen replied:
and do you mean to ask Marc or is it `mac`?

OlegHahm replied:
It's indeed called `MARCSTATE` in the cc11xx (MAin Radio Control State Machine State).

kaspar030 replied:
I just removed all of those functions, they were unused anyways.

haukepetersen commented on boards/msba2/include/cc110x_params.h:
```diff
@@ -0,0 +1,44 @@
+const cc110x_params_t cc110x_params[] = {
+    {
+        .spi  = 0,
+        .CS   = 53,
```
why are these capitalized?

haukepetersen commented on drivers/cc110x/cc110x-netdev2.c:
```diff
@@ -0,0 +1,218 @@
+    DEBUG("%s:%u\n", __func__, __LINE__);
+
+    netdev2_cc110x_t *netdev2_cc110x = (netdev2_cc110x_t*) dev;
+    cc110x_pkt_t *cc110x_pkt = vector[0].iov_base;
```
do you only send the first entry from the ivec list?

kaspar030 replied:
Yes, the driver currently accepts only that. But gnrc_cc110x converts a full pktsnip into one iovec.

haukepetersen replied:
I see

haukepetersen commented on drivers/cc110x/cc110x-spi.c:
```diff
@@ -0,0 +1,175 @@
+{
+    unsigned int cpsr;
+    spi_acquire(dev->params.spi);
+    cpsr = disableIRQ();
```
why are you disabling interrupts during transfer? The transfer should be thread safe, so no need?!

kaspar030 replied:
I did take that over from the old implementation. It seemed weird to me, too, but maybe without disabling interrupts, the whole driver's logic breaks. Let's investigate later.

haukepetersen replied:
check

haukepetersen commented on drivers/cc110x/cc110x.c:
```diff
@@ -0,0 +1,393 @@
+    gpio_init(dev->params.GDO1, GPIO_DIR_IN, GPIO_NOPULL);
+
+    /* Configure SPI */
+    spi_poweron(dev->params.spi);
```
no need, spi is powered on on `spi_init_xx()`

haukepetersen commented on drivers/include/cc110x.h:
```diff
@@ -0,0 +1,128 @@
+    gpio_t GDO2;
+} cc110x_params_t;
+
+typedef struct cc110x cc110x_t;
```
you could remove this and put the name of the struct into the definition below..

kaspar030 replied:
No, as struct cc110x itself uses cc110x_t, the forward declaration is needed.

haukepetersen commented on drivers/include/cc110x.h:
```diff
@@ -0,0 +1,128 @@
+
+/**
+ * @defgroup    drivers_cc110x CC110x
+ * @ingroup     drivers_cc110x
```
ingroup `drivers`

haukepetersen replied:
also `@brief` missing

haukepetersen commented on drivers/include/cc110x/cc110x-interface.h:
```diff
@@ -0,0 +1,62 @@
+
+/**
+ * @defgroup    drivers_cc110x CC110X
+ * @brief       Driver for Texas Instruments CC110x (without MAC protocol)
```
the the without MAC still true?

haukepetersen replied:
also just leave the `@ingroup drivers_cc110x` and remove the rest

haukepetersen commented on drivers/include/cc110x/cc110x-internal.h:
```diff
@@ -0,0 +1,202 @@
+/**
+ * @defgroup    drivers_cc110x CC110X
+ * @brief       Driver for Texas Instruments CC110x (without MAC protocol)
+ * @ingroup     drivers
```
duplicate group definition -> just leave the ingroup and remove the rest

haukepetersen commented on drivers/include/cc110x/cc110x-internal.h:
```diff
@@ -0,0 +1,202 @@
+/*
```
might make sense to joing the xx-internal.h and the xx-defines.h files...

haukepetersen replied:
ah, forget about it - when looking at the internal structure of the driver the separation doesn't hurt...

haukepetersen commented on drivers/include/cc110x/cc110x-netdev2.h:
```diff
@@ -0,0 +1,52 @@
+ */
+typedef struct netdev2_cc110x {
+    netdev2_t netdev;
+    cc110x_t cc110x;
```
doxygen will complain here, too

OlegHahm commented:
@kaspar030, planning to address the comments?

kaspar030 commented:
@OlegHahm Yes, but I want to finish #3525 first.

OlegHahm commented:
Sure, just checking.

kaspar030 commented:
- addressed most comments
- rebased

The gnrc part of this drivers doesn't play well, yet, with sixlowpan. The crude adaption layer needs some work.

For the node ID, what's the current state of the art for getting a random or node-unique number?

kaspar030 commented:
*update* When disabling IPHC, together with #3944, I can ping between two msba2's.

@authmillenon Do you know any reason why IPHC might break? I suspect it's because of the 1 byte hardware addresses...

authmillenon commented:
> @authmillenon Do you know any reason why IPHC might break? I suspect it's because of the 1 byte hardware addresses...

I took special care that it would work with 1-byte addresses, but never tested it (duh). So there might still be some hidden issues.

kaspar030 commented:
- squashed

kaspar030 commented:
- waiting for #3946.

haukepetersen commented:
not anymore...

haukepetersen commented:
kicked Travis

OlegHahm commented:
Do we have an ACK somewhere?

haukepetersen commented:
let me look over this once more, then you'll have it!

haukepetersen commented:
Looks good. ACK and go.
